---
title: "G27 Genome Search"
author: "Ryan Johnson"
date: "1/28/2022"
output: html_document
---

```{r message=FALSE, warning=FALSE}
# Load packages
library(biomartr) # Be sure to download dev version 
library(Biostrings)
library(tidyverse)
library(gt)
library(WriteXLS)

# Set timeout to avoid query issue
options(timeout = 30000)
```


# Retrieve G27 Sequence

```{r message=FALSE, warning=FALSE}
# Pull in genbank sequence
g27_genbank <- biomartr::getGenome(db = "genbank",
                                   organism = "Helicobacter pylori G27",
                                   path = "G27_genbank/",
                                   reference = FALSE)

# Get GFF file
g27_gff_path <-  getGFF(db = "genbank",
                          organism = "Helicobacter pylori G27",
                          reference = FALSE,
                          release = NULL,
                          gunzip = FALSE,
                          path = file.path("G27_genbank/annotation"))

# Read genome
g27_seq <- biomartr::read_genome(file = g27_genbank, format = "fasta")

# Read gff
g27_gff <- biomartr::read_gff(file = g27_gff_path)
```

# Create annotation table

```{r}
# Create function
hits_in_genome <- function(fasta, GFF, nuc_seq){
  
  # Extract seq match coordinates on genome
  match_index <- vmatchPattern(nuc_seq, fasta)
  start_index <- startIndex(match_index)[[1]]
  end_index <- endIndex(match_index)[[1]]
  
  # Filter GFF for just CDS regions
  gff_cds <- GFF %>% 
    filter(type == "CDS")
  
  # Initialize empty results tibble
  res_tib <- tibble(hit_start = numeric(),
                    hit_end = numeric(),
                    CDS_start = numeric(),
                    CDS_end = numeric(),
                    Strand = character(),
                    Attribute = character())
  
  # Iterate through start index
  for(i in start_index){
    query <- gff_cds %>% 
      filter(i >= start & i <= end)
    
    # If in CDS
    if(nrow(query) > 0){
      CDS_hit <- tibble(hit_start = i,
                        hit_end = end_index[start_index == i],
                        CDS_start = query$start,
                        CDS_end = query$end,
                        Strand = query$strand,
                        Attribute = query$attribute)
      
       res_tib <- res_tib %>% 
         full_join(CDS_hit, by = c("hit_start", "hit_end", "CDS_start", 
                                   "CDS_end", "Strand", "Attribute"))
    }
    
    # If between CDS
    else if(nrow(query) == 0){
      CDS_miss <- tibble(hit_start = i,
                         hit_end = end_index[start_index == i],
                         CDS_start = NULL,
                         CDS_end = NULL,
                         Strand = NULL,
                         Attribute = "IGR")
      
      res_tib <- res_tib %>% 
         full_join(CDS_miss, by = c("hit_start", "hit_end", "Attribute"))
    }
  }
  res_tib
}
```


### Hits in Chromosome

```{r message=FALSE, warning=FALSE}
# Filter gff file for just chromosome
g27_gff_chr <- g27_gff %>% 
  filter(!grepl('HPG27_A', attribute)) %>% 
  filter(!grepl("CP001174.1", attribute))

chr_hits_f <- hits_in_genome(fasta = g27_seq[1],
                           GFF = g27_gff_chr, 
                           nuc_seq = "ATTAAT") %>% 
  mutate(Chr_or_Plsmd = "Chr") %>% 
  mutate(query_seq = "ATTAAT")

chr_hits_r <- hits_in_genome(fasta = g27_seq[1],
                           GFF = g27_gff_chr, 
                           nuc_seq = "TAATTA") %>% 
  mutate(Chr_or_Plsmd = "Chr") %>% 
  mutate(query_seq = "TAATTA")
```

### Hits in Plasmid

### Hits in Chromosome

```{r message=FALSE, warning=FALSE}
# Filter gff file for just plasmid
g27_gff_pls <- g27_gff %>% 
  filter(grepl('HPG27_A', attribute))

pls_hits_f <- hits_in_genome(fasta = g27_seq[2],
                           GFF = g27_gff_pls, 
                           nuc_seq = "ATTAAT") %>% 
  mutate(Chr_or_Plsmd = "Plsmd") %>% 
  mutate(query_seq = "ATTAAT")

pls_hits_r <- hits_in_genome(fasta = g27_seq[2],
                           GFF = g27_gff_pls, 
                           nuc_seq = "TAATTA") %>% 
  mutate(Chr_or_Plsmd = "Plsmd") %>% 
  mutate(query_seq = "TAATTA")
```


### Merge All

```{r}
all_hits <- chr_hits_f %>% 
  full_join(chr_hits_r) %>% 
  full_join(pls_hits_f) %>% 
  full_join(pls_hits_r)

write_excel_csv(all_hits, "all_hits_g27.csv")
```














